### PR TITLE
Re-implement tf old-style extractors with extractor extensions

### PR DESCRIPTION
**Pull request by evolosen**
Ticket: #44806

Added ConcatV2, FusedBatchNorm, FusedBatchNormV2, FusedBatchNormV3, Pack, RandomUniform extractors as class extractors. Added BatchNormTraining operation and deleted FusedBatchNormNonConstant transformation. Fixed MVNReplacer for 'NCHW'.

Code:
* [x]  Comments
* [x]  Code style (PEP8): N/A
* [x]  Transformation generates reshape-able IR: N/A
* [x]  Transformation preserves original framework node names

Validation:
* [x]  Unit tests: Modified FusedBatchNormTraining_test.py, added tf/mvn_test.py
* [x]  Framework operation tests: N/A
* [x]  Transformation tests: N/A
* [x]  e2e model test with an update of ./tests/e2e_oss/utils/reshape_utils.py: N/A
* [x]  Model Optimizer IR Reader check: N/A

Documentation:
* [x]  Supported frameworks operations list: N/A
* [x]  Supported **public** models list: N/A
* [x]  New operations specification: N/A
* [x]  Guide on how to convert the **public** model: N/A
* [x]  User guide update: N/A

### CODE REVIEW

File: model-optimizer/extensions/front/tf/pack_ext.py [hunk 1]
```diff
@@ -13,11 +13,23 @@
  See the License for the specific language governing permissions and
  limitations under the License.
 """
+from extensions.ops.pack import PackOp
+from mo.front.extractor import FrontExtractorOp
+from mo.front.tf.extractors.utils import tf_dtype_extractor
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/front/tf/pack_ext.py:19]: Remove empty line

File: model-optimizer/extensions/middle/FusedBatchNormNonConstant.py [hunk 2]
```diff
@@ -40,8 +40,7 @@ def run_before(self):
     def pattern(self):
         return dict(
             nodes=[
-                ('op', dict(kind='op', op=lambda op: op in ['FusedBatchNorm', 'FusedBatchNormV2',
-                                                            'FusedBatchNormV3']))],
+                ('op', dict(kind='op', op=lambda op: op in ['batchNormInference']))],
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/middle/FusedBatchNormNonConstant.py:43]: Change to `('op', dict(kind='op', op='batchNormInference'))],`

File: model-optimizer/extensions/ops/RandomUniformOp.py [hunk 3]
```diff
@@ -13,14 +13,21 @@
  See the License for the specific language governing permissions and
  limitations under the License.
 """
+from mo.graph.graph import Graph
+from mo.ops.op import Op
+from mo.front.common.partial_infer.random_uniform import tf_random_uniform_infer
 
-from mo.front.common.partial_infer.concat import tf_pack_infer
+class RandomUniformOp(Op):
+    op = 'RandomUniformOp'
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/ops/RandomUniformOp.py:21]: Remove 'Op' from the name of the class, file and attribute "op"

File: model-optimizer/extensions/ops/RandomUniformOp.py [hunk 4]
```diff
@@ -13,14 +13,21 @@
  See the License for the specific language governing permissions and
  limitations under the License.
 """
+from mo.graph.graph import Graph
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/ops/RandomUniformOp.py:16]: sort `imports`

File: model-optimizer/extensions/ops/RandomUniformOp.py [hunk 5]
```diff
@@ -13,14 +13,21 @@
...
+from mo.ops.op import Op
+from mo.front.common.partial_infer.random_uniform import tf_random_uniform_infer
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/ops/RandomUniformOp.py:18]: Remove file `mo/front/common/partial_infer/random_uniform.py` and use `copy_shape_infer` method.

File: model-optimizer/mo/front/tf/extractor.py [hunk 6]
```diff
@@ -14,11 +14,7 @@
  limitations under the License.
 """
 
-from mo.front.tf.extractors.concat import tf_concat_ext
-from mo.front.tf.extractors.fused_bn import tf_fused_bn_extractor
 from mo.front.tf.extractors.native_tf import native_tf_node_extractor
-from mo.front.tf.extractors.pack import tf_pack_ext
-from mo.front.tf.extractors.random_uniform import tf_random_uniform_ext
 from mo.front.tf.extractors.utils import get_tf_node_port
```
Comment (lazarevevgeny) [inline model-optimizer/mo/front/tf/extractor.py:18]: The `get_tf_node_port` function is not used in this file

File: model-optimizer/mo/middle/passes/fusing/decomposition.py [hunk 7]
```diff
@@ -39,8 +39,8 @@ def convert_batch_norm(graph: Graph):
     """
     nodes = graph.get_op_nodes()
     for node in nodes:
-        if node.has_valid('op') and (node.op in ['FusedBatchNorm', 'FusedBatchNormV2', 'FusedBatchNormV3',
-                                                 'BatchNorm', 'BatchNormalization', 'batchNormInference']):
+        if node.has_valid('op') and (node.op in ['BatchNorm', 'BatchNormalization', 
+                                        'batchNormInference']):
```
Comment (lazarevevgeny) [inline model-optimizer/mo/middle/passes/fusing/decomposition.py:43]: nit: Merge into one line. It should fit into 120 characters limit.

**Review by lazarevevgeny — CHANGES_REQUESTED** [review]

File: model-optimizer/extensions/middle/FusedBatchNormTraining.py [hunk 8]
```diff
@@ -44,15 +46,14 @@ class FusedBatchNormTraining(MiddleReplacementPattern):
     def pattern(self):
         return dict(
             nodes=[
-                ('op', dict(kind='op', op=lambda op: op in ['FusedBatchNorm', 'FusedBatchNormV2', 'FusedBatchNormV3'],
-                 is_training=True))],
+                ('op', dict(kind='op', op=lambda op: op in ['batchNormTraining']))],
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/middle/FusedBatchNormTraining.py:49]: Change to `('op', dict(kind='op', op='batchNormTraining'))],`

File: model-optimizer/extensions/middle/FusedBatchNormTraining.py [hunk 9]
```diff
@@ -44,15 +46,14 @@ class FusedBatchNormTraining(MiddleReplacementPattern):
...
             edges=[]
         )
 
     def replace_pattern(self, graph: Graph, match: dict):
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/middle/FusedBatchNormTraining.py:53]: Refactor this transformation to the following approach (and remove `pattern` function):

```python
    def find_and_replace_pattern(self, graph: Graph):
        for batch_norm_training_node in graph.get_op_nodes(op='batchNormTraining'):
```

File: model-optimizer/extensions/middle/FusedBatchNormTraining.py [hunk 10]
```diff
@@ -81,3 +82,14 @@ def replace_pattern(self, graph: Graph, match: dict):
...
         reshape_back.in_port(1).connect(original_shape.out_port(0))
         mvn.out_port(0).get_connection().insert_node(reshape_back)
+
+        @staticmethod
+        def castBatchNormTraining2Inference(batchNormTrainingNode: Node) -> Node:
+            additional_attrs = {}
+            for batchNormAttr in  batchNormAttrList:
+                if batchNormTrainingNode.has(batchNormAttr):
+                    additional_attrs[batchNormAttr] = batchNormTrainingNode[batchNormAttr]
+            additional_attrs['name'] = batchNormTrainingNode.name
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/middle/FusedBatchNormTraining.py:92]: Use `rename_nodes` function to rename original batchNormTraining node with batchNormInference node.

File: model-optimizer/extensions/middle/FusedBatchNormTraining.py [hunk 11]
```diff
@@ -81,3 +82,14 @@ def replace_pattern(self, graph: Graph, match: dict):
                                        'override_output_shape': True}).create_node()
         reshape_back.in_port(1).connect(original_shape.out_port(0))
         mvn.out_port(0).get_connection().insert_node(reshape_back)
+
+        @staticmethod
+        def castBatchNormTraining2Inference(batchNormTrainingNode: Node) -> Node:
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/middle/FusedBatchNormTraining.py:87]: Merge this function into the main transformation function.

File: model-optimizer/extensions/middle/RandomUniformReplacer.py [hunk 12]
```diff
@@ -39,7 +39,7 @@ def pattern():
             nodes=[
                 ('shape', dict(op='ShapeOf')),
                 ('shape_data', dict()),
-                ('random_uniform', dict(op='RandomUniform')),
+                ('random_uniform', dict(op='RandomUniformOp')),
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/middle/RandomUniformReplacer.py:42]: Revert this change

File: model-optimizer/extensions/ops/BatchNormInference.py [hunk 13]
```diff
@@ -16,6 +16,7 @@
 
 from mo.graph.graph import Graph
 from mo.ops.op import Op
+import numpy as np
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/ops/BatchNormInference.py:19]: Sort imports

File: model-optimizer/extensions/middle/RandomUniformReplacer_test.py [hunk 14]
```diff
@@ -24,7 +24,7 @@
 nodes_attributes = {
     'shape': {'kind': 'op', 'op': 'ShapeOf'},
     'shape_data': {'kind': 'data'},
-    'random_uniform': {'kind': 'op', 'op': 'RandomUniform'},
+    'random_uniform': {'kind': 'op', 'op': 'RandomUniformOp'},
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/middle/RandomUniformReplacer_test.py:27]: Revert this change.

File: model-optimizer/extensions/ops/BatchNormInference.py [hunk 15]
```diff
@@ -36,4 +37,6 @@ def __init__(self, graph: Graph, attrs: dict):
         }, attrs)
     @staticmethod
     def infer(node):
-        node.out_port(0).data.set_shape(node.in_port(0).data.get_shape())
+        output_shape = np.array(node.in_node(0).shape)
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/ops/BatchNormInference.py:40]: Batch norm inference has just one output. So the change is invalid. You should use `copy_shape_infer` here instead of custom shape infer function.

**Review by lazarevevgeny — CHANGES_REQUESTED** [review]

File: model-optimizer/extensions/middle/FusedBatchNormTraining.py [hunk 16]
```diff
@@ -44,15 +46,14 @@ class FusedBatchNormTraining(MiddleReplacementPattern):
...
             edges=[]
         )
 
     def replace_pattern(self, graph: Graph, match: dict):
-        node = match['op']
-        node.is_training = False
-
+        old_node = match['op']
+        node = castBatchNormTraining2Inference(old_node)
+        old_node.replace_node(node)
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/middle/FusedBatchNormTraining.py:56]: Please, do not use `replace_node` method. Instead, create a new `batchNormInference` node and connect data to its inputs (5) and **just one output**.

File: model-optimizer/extensions/ops/BatchNormTraining.py [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+from mo.ops.op import Op
+import numpy as np
+
+
+class BatchNormTraining(Op):
+    """
+    BatchNormInference will be replaced by BatchNormInference after 
+        FusedBatchNormTraining tranformation 
+    """
+    op = 'batchNormTraining'
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/ops/BatchNormTraining.py:27]: Let's change the first character to capital 'B'. This is our standard approach. The "batchNormInference" is some strange exception.

File: model-optimizer/extensions/middle/FusedBatchNormNonConstant.py [hunk 18]
```diff
@@ -40,8 +40,7 @@ def run_before(self):
...
             edges=[]
         )
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/middle/FusedBatchNormNonConstant.py:46]: We can modify this transformation to be executed during the front phase and no matter that inputs are non constants. All the values will be const-folded automatically later in the pipeline. Note, that methods like `create_node_with_data` should not be used. Instead, `Port`s and `Connection`s should be used.

File: model-optimizer/mo/middle/passes/fusing/decomposition.py [hunk 19]
```diff
@@ -39,8 +39,8 @@ def convert_batch_norm(graph: Graph):
     """
     nodes = graph.get_op_nodes()
     for node in nodes:
-        if node.has_valid('op') and (node.op in ['FusedBatchNorm', 'FusedBatchNormV2', 'FusedBatchNormV3',
```
Comment (lazarevevgeny) [inline model-optimizer/mo/middle/passes/fusing/decomposition.py:42]: What is the difference between `model-optimizer/extensions/middle/FusedBatchNormNonConstant.py` and this function? Can we use just one of them?

**Review by lazarevevgeny — CHANGES_REQUESTED** [review]

File: model-optimizer/extensions/middle/RandomUniformReplacer.py (hunk 12 above)
Comment (evolosen) [reply model-optimizer/extensions/middle/RandomUniformReplacer.py]: fixed

File: model-optimizer/extensions/middle/RandomUniformReplacer_test.py (hunk 14 above)
Comment (evolosen) [reply model-optimizer/extensions/middle/RandomUniformReplacer_test.py]: done

File: model-optimizer/extensions/front/tf/fused_bn_ext.py [hunk 20]
```diff
@@ -0,0 +1,64 @@
...
+        'eps': pb.attr['epsilon'].f
+    }
+    (BatchNormTraining if is_training else BatchNormInference)\
+    .update_node_stat(node, attrs)
+
+
+class FusedBatchNormExtractor(FrontExtractorOp):
+    op = "FusedBatchNorm"
+    enabled = True
+    @classmethod
```
[the remark this answers is not in the document]
Comment (evolosen) [reply model-optimizer/extensions/front/tf/fused_bn_ext.py:43]: done

File: model-optimizer/extensions/front/tf/pack_ext.py (hunk 1 above)
Comment (evolosen) [reply model-optimizer/extensions/front/tf/pack_ext.py:19]: done

File: model-optimizer/extensions/middle/FusedBatchNormNonConstant.py (hunk 2 above)
Comment (evolosen) [reply model-optimizer/extensions/middle/FusedBatchNormNonConstant.py:43]: FusedBatchNormNonConstant.py  was deleted

File: model-optimizer/extensions/middle/FusedBatchNormNonConstant.py (hunk 18 above)
Comment (evolosen) [reply model-optimizer/extensions/middle/FusedBatchNormNonConstant.py:46]: No, these transforms depend on another  middle phase transforms on some models.

File: model-optimizer/extensions/ops/BatchNormInference.py (hunk 15 above)
Comment (evolosen) [reply model-optimizer/extensions/ops/BatchNormInference.py:40]: copy_shape_infer contains additional checks for value. Do we need this?

File: model-optimizer/extensions/ops/BatchNormInference.py (hunk 13 above)
Comment (evolosen) [reply model-optimizer/extensions/ops/BatchNormInference.py:19]: What sorting rule is?

File: model-optimizer/extensions/ops/BatchNormTraining.py [hunk 21]
```diff
@@ -0,0 +1,42 @@
...
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+"""
+
+from mo.graph.graph import Graph
+from mo.ops.op import Op
```
[the remark this answers is not in the document]
Comment (evolosen) [reply model-optimizer/extensions/ops/BatchNormTraining.py:18]: What sorting rule is?

File: model-optimizer/extensions/ops/BatchNormTraining.py [hunk 22]
```diff
@@ -0,0 +1,42 @@
...
+import numpy as np
+
+
+class BatchNormTraining(Op):
+    """
+    BatchNormInference will be replaced by BatchNormInference after 
+        FusedBatchNormTraining tranformation
```
[the remark this answers is not in the document]
Comment (evolosen) [reply model-optimizer/extensions/ops/BatchNormTraining.py:25]: fixed

File: model-optimizer/extensions/ops/BatchNormTraining.py [hunk 23]
```diff
@@ -0,0 +1,42 @@
...
+    """
+    op = 'batchNormTraining'
+    enabled = False
+
+    def __init__(self, graph: Graph, attrs: dict):
+        super().__init__(graph, {
+            'type': None,
+            'op': self.op,
+            'in_ports_count': 5,
+            'out_ports_count': 1,
```
[the remark this answers is not in the document]
Comment (evolosen) [reply model-optimizer/extensions/ops/BatchNormTraining.py:35]: fixed

File: model-optimizer/extensions/ops/BatchNormTraining.py [hunk 24]
```diff
@@ -0,0 +1,42 @@
...
+            'infer': self.infer
```
[the remark this answers is not in the document]
Comment (evolosen) [reply model-optimizer/extensions/ops/BatchNormTraining.py:36]: We cant do it.

File: model-optimizer/mo/front/tf/extractor.py (hunk 6 above)
Comment (evolosen) [reply model-optimizer/mo/front/tf/extractor.py:18]: src_node, src_port = get_tf_node_port(src_node_id)   --  32 line

File: model-optimizer/mo/middle/passes/fusing/decomposition.py (hunk 19 above)
Comment (evolosen) [reply model-optimizer/mo/middle/passes/fusing/decomposition.py:42]: fixed

File: model-optimizer/mo/middle/passes/fusing/decomposition.py (hunk 7 above)
Comment (evolosen) [reply model-optimizer/mo/middle/passes/fusing/decomposition.py:43]: fixed

File: model-optimizer/extensions/front/tf/fused_bn_ext.py [hunk 25]
```diff
@@ -0,0 +1,64 @@
...
+def tf_fused_batch_norm_extract(cls, node: Node):
+    pb = node.pb
+    is_training = pb.attr['is_training'].b
+    attrs = {
+        'data_format': pb.attr["data_format"].s,
+        'data_type': tf_dtype_extractor(pb.attr["T"].type),
+        'eps': pb.attr['epsilon'].f
+    }
+    (BatchNormTraining if is_training else BatchNormInference)\
+    .update_node_stat(node, attrs)
```
[the remark this answers is not in the document]
Comment (evolosen) [reply model-optimizer/extensions/front/tf/fused_bn_ext.py:37]: fixed

File: model-optimizer/extensions/front/tf/fused_bn_ext.py [hunk 26]
```diff
@@ -0,0 +1,64 @@
...
+
+from mo.front.extractor import FrontExtractorOp
+from mo.graph.graph import Node
+from extensions.ops.BatchNormInference import BatchNormInference
+from extensions.ops.BatchNormTraining import BatchNormTraining
+from mo.front.tf.common import tf_data_type_decode
+
+def tf_dtype_extractor(pb_dtype, default=None):
+    return tf_data_type_decode[pb_dtype][0] if pb_dtype in tf_data_type_decode else default
+
```
[the remark this answers is not in the document]
Comment (evolosen) [reply model-optimizer/extensions/front/tf/fused_bn_ext.py:25]: fixed

File: model-optimizer/extensions/ops/elementwise.py [hunk 27]
```diff
@@ -1,5 +1,5 @@
 """
- Copyright (C) 2018-2020 Intel Corporation
+ Copyright (C) 2018-2021 Intel Corporation
```
Comment (rkazants) [inline model-optimizer/extensions/ops/elementwise.py:2]: Why is copyright changed if nothing in the code is touched?

File: model-optimizer/mo/front/common/partial_infer/concat.py [hunk 28]
```diff
@@ -26,7 +26,7 @@
 
 
 def concat_infer(node):
-    if not node.has('axis'):
+    if not node.has('axis') or node.axis is None:
```
Comment (rkazants) [inline model-optimizer/mo/front/common/partial_infer/concat.py:29]: change copyright.

File: model-optimizer/extensions/front/tf/mvn.py [hunk 29]
```diff
@@ -35,7 +35,7 @@ def pattern(self):
                 ('variance', dict(op='ReduceMean')),
                 ('squeeze_mean', dict(op='Squeeze')),
                 ('squeeze_variance', dict(op='Squeeze')),
-                ('fbn', dict(op=lambda op: op in ['FusedBatchNorm', 'FusedBatchNormV2', 'FusedBatchNormV3'])),
+                ('fbn', dict(op=lambda op: op in ['batchNormInference'])),
```
Comment (rkazants) [inline model-optimizer/extensions/front/tf/mvn.py:38]: I think `batchNormInference` must be capitalized

**Review by vgavrilo — APPROVED** [review]

File: model-optimizer/extensions/front/tf/random_uniform_ext.py [hunk 30]
```diff
@@ -15,9 +15,15 @@
 """
 
 from mo.front.common.partial_infer.random_uniform import tf_random_uniform_infer
+from mo.front.extractor import FrontExtractorOp
+from mo.ops.concat import Concat
+from extensions.ops.RandomUniform import RandomUniform
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/front/tf/random_uniform_ext.py:20]: Sort imports and update copyright year

File: model-optimizer/extensions/ops/BatchNormTraining.py (hunk 17 above)
Comment (lazarevevgeny) [reply model-optimizer/extensions/ops/BatchNormTraining.py:27]: Change to capital `B`.

File: model-optimizer/extensions/middle/FusedBatchNormTraining.py (hunk 8 above)
Comment (lazarevevgeny) [reply model-optimizer/extensions/middle/FusedBatchNormTraining.py:51]: My original comment was wrong. Sorry for that. The `B` must be capital.

File: model-optimizer/extensions/middle/FusedBatchNormTraining.py [hunk 31]
```diff
@@ -44,14 +48,33 @@ class FusedBatchNormTraining(MiddleReplacementPattern):
...
-        node.is_training = False
+        bn_train_node = match['op']
+        additional_attrs = {}
+        for batchNormAttr in batchNormAttrList:
+            if bn_train_node.has(batchNormAttr):
+                additional_attrs[batchNormAttr] = bn_train_node[batchNormAttr]
+
+        if len(bn_train_node.out_nodes().items()) > 1:
+            additional_attrs['name'] = bn_train_node.name + '/batchNormInferenceMO'
+            node = BatchNormInference(bn_train_node.graph, additional_attrs).create_node()
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/middle/FusedBatchNormTraining.py:64]: May be you should use `BatchNormInferenceMO` here?

File: model-optimizer/extensions/middle/FusedBatchNormTraining.py [hunk 32]
```diff
@@ -44,14 +48,33 @@ class FusedBatchNormTraining(MiddleReplacementPattern):
...
+            for port_id, out_node in bn_train_node.out_nodes().items():
+                bn_train_node.out_port(port_id).get_connection().set_source(node.out_port(port_id))
+
+        elif len(bn_train_node.out_nodes().items()) == 1:
+            additional_attrs['name'] = bn_train_node.name + '/batchNormInference'
+            node = BatchNormInference(bn_train_node.graph, additional_attrs).create_node()
+            bn_train_node.out_port(0).get_connection().set_source(node.out_port(0))
+
+        else:
+            assert False, 'Node  {} has not out nodes'.format(bn_train_node.name)
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/middle/FusedBatchNormTraining.py:74]: nit: `'Node {} does not have output nodes'`

File: model-optimizer/extensions/middle/FusedBatchNormTraining.py [hunk 33]
```diff
@@ -44,14 +48,33 @@ class FusedBatchNormTraining(MiddleReplacementPattern):
...
-        node = match['op']
-        node.is_training = False
+        bn_train_node = match['op']
+        additional_attrs = {}
+        for batchNormAttr in batchNormAttrList:
+            if bn_train_node.has(batchNormAttr):
+                additional_attrs[batchNormAttr] = bn_train_node[batchNormAttr]
+
+        if len(bn_train_node.out_nodes().items()) > 1:
+            additional_attrs['name'] = bn_train_node.name + '/batchNormInferenceMO'
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/middle/FusedBatchNormTraining.py:63]: Do not use `bn_train_node.name`. Instead create a variable `bn_train_node_name = bn_train_node.soft_get('name', bn_train_node.id)` and use it everywhere in the code

File: model-optimizer/extensions/ops/BatchNormInferenceMultipleOutputs.py [hunk 34]
```diff
@@ -0,0 +1,42 @@
...
+"""
+
+from mo.graph.graph import Graph
+from mo.ops.op import Op
+import numpy as np
+
+
+class BatchNormInferenceMO(Op):
+    """
+    BatchNormInference with Multiple outputs
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/ops/BatchNormInferenceMultipleOutputs.py:24]: Please, extend the description of this operation. I think it make sense to add that this operation is used to extract TensorFlow FusedBatchNorm operation.

File: model-optimizer/extensions/ops/BatchNormInferenceMultipleOutputs.py [hunk 35]
```diff
@@ -0,0 +1,42 @@
...
+            'type': None,
+            'op': self.op,
+            'in_ports_count': 5,
+            'out_ports_count': 6,
+            'infer': self.infer
+        }, attrs)
+
+    @staticmethod
+    def infer(node):
+        output_shape = np.array(node.in_node(0).shape)
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/ops/BatchNormInferenceMultipleOutputs.py:40]: Change to `output_shape = int64_array(node.in_port(0).data.get_shape())`

File: model-optimizer/extensions/ops/BatchNormInferenceMultipleOutputs.py [hunk 36]
```diff
@@ -0,0 +1,42 @@
...
+        for port, out_node in node.out_nodes().items():
+            out_node.shape = output_shape
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/ops/BatchNormInferenceMultipleOutputs.py:42]: Change to 
```py
for out_port in node.out_ports().values():
    out_port.data.set_shape(output_shape)
```

File: model-optimizer/extensions/ops/BatchNormTraining.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (lazarevevgeny) [reply model-optimizer/extensions/ops/BatchNormTraining.py:18]: Use IDE feature to sort them.

File: model-optimizer/extensions/ops/BatchNormTraining.py [hunk 37]
```diff
@@ -0,0 +1,43 @@
...
+            'in_ports_count': 5,
+            'out_ports_count': 6,
+            'infer': self.infer
+        }, attrs)
+
+    @staticmethod
+    def infer(node):
+        output_shape = np.array(node.in_node(0).shape)
+        for port, out_node in node.out_nodes().items():
+            out_node.shape = output_shape
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/ops/BatchNormTraining.py:43]: Refactor like in the `BatchNormInferenceMO` operation

File: model-optimizer/extensions/ops/RandomUniform.py [hunk 38]
```diff
@@ -13,13 +13,21 @@
  See the License for the specific language governing permissions and
  limitations under the License.
 """
+from mo.graph.graph import Graph
+from mo.ops.op import Op
+from mo.front.common.partial_infer.random_uniform import tf_random_uniform_infer
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/ops/RandomUniform.py:18]: sort imports

File: model-optimizer/extensions/ops/RandomUniform.py [hunk 39]
```diff
@@ -13,13 +13,21 @@
...
 
-from mo.front.common.partial_infer.concat import concat_infer
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/ops/RandomUniform.py:17]: Add one more empty line before the class

**Review by lazarevevgeny — CHANGES_REQUESTED** [review]

File: model-optimizer/extensions/middle/FusedBatchNormTraining.py (hunk 31 above)
Comment (evolosen) [reply model-optimizer/extensions/middle/FusedBatchNormTraining.py:64]: fixed

File: model-optimizer/extensions/front/caffe/bn.py [hunk 40]
```diff
@@ -27,7 +27,7 @@ class BNToScaleShift(FrontReplacementOp):
     """
     Replaces BN layer with ScaleShift.
```
[the remark this answers is not in the document]
Comment (evolosen) [reply model-optimizer/extensions/front/caffe/bn.py:28]: fixed

File: model-optimizer/extensions/front/tf/concat_v2_ext.py [hunk 41]
```diff
@@ -14,10 +14,15 @@
  limitations under the License.
 """
 
-from mo.front.common.partial_infer.elemental import copy_shape_infer
```
[the remark this answers is not in the document]
Comment (evolosen) [reply model-optimizer/extensions/front/tf/concat_v2_ext.py:17]: done

File: model-optimizer/extensions/front/tf/fused_bn_ext.py [hunk 42]
```diff
@@ -0,0 +1,67 @@
...
+ See the License for the specific language governing permissions and
+ limitations under the License.
+"""
+
+from mo.front.extractor import FrontExtractorOp
+from mo.graph.graph import Node
+from extensions.ops.BatchNormInference import BatchNormInference
+from extensions.ops.BatchNormInferenceMultipleOutputs import BatchNormInferenceMO
+from extensions.ops.BatchNormTraining import BatchNormTraining
+from mo.front.tf.extractors.utils import tf_dtype_extractor
```
[the remark this answers is not in the document]
Comment (evolosen) [reply model-optimizer/extensions/front/tf/fused_bn_ext.py:22]: done

File: model-optimizer/extensions/front/tf/mvn.py (hunk 29 above)
Comment (evolosen) [reply model-optimizer/extensions/front/tf/mvn.py:38]: done

File: model-optimizer/extensions/front/tf/random_uniform_ext.py (hunk 30 above)
Comment (evolosen) [reply model-optimizer/extensions/front/tf/random_uniform_ext.py:20]: done

File: model-optimizer/extensions/middle/FusedBatchNormTraining.py (hunk 33 above)
Comment (evolosen) [reply model-optimizer/extensions/middle/FusedBatchNormTraining.py:63]: done

File: model-optimizer/extensions/ops/BatchNormInferenceMultipleOutputs.py (hunk 34 above)
Comment (evolosen) [reply model-optimizer/extensions/ops/BatchNormInferenceMultipleOutputs.py:24]: done

File: model-optimizer/extensions/ops/BatchNormInferenceMultipleOutputs.py (hunk 35 above)
Comment (evolosen) [reply model-optimizer/extensions/ops/BatchNormInferenceMultipleOutputs.py:40]: done

File: model-optimizer/extensions/ops/BatchNormInferenceMultipleOutputs.py (hunk 36 above)
Comment (evolosen) [reply model-optimizer/extensions/ops/BatchNormInferenceMultipleOutputs.py:42]: done

File: model-optimizer/extensions/ops/BatchNormTraining.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (evolosen) [reply model-optimizer/extensions/ops/BatchNormTraining.py:18]: done

File: model-optimizer/extensions/ops/BatchNormTraining.py (hunk 17 above)
Comment (evolosen) [reply model-optimizer/extensions/ops/BatchNormTraining.py:27]: done

File: model-optimizer/extensions/ops/BatchNormTraining.py (hunk 37 above)
Comment (evolosen) [reply model-optimizer/extensions/ops/BatchNormTraining.py:43]: done

File: model-optimizer/extensions/ops/RandomUniform.py (hunk 38 above)
Comment (evolosen) [reply model-optimizer/extensions/ops/RandomUniform.py:18]: done

File: model-optimizer/extensions/ops/RandomUniform.py (hunk 39 above)
Comment (evolosen) [reply model-optimizer/extensions/ops/RandomUniform.py:17]: done

File: model-optimizer/mo/front/common/partial_infer/concat.py (hunk 28 above)
Comment (evolosen) [reply model-optimizer/mo/front/common/partial_infer/concat.py:29]: done

**Review by vgavrilo — APPROVED** [review]

**Review by vgavrilo — APPROVED** [review]

**Comment by rkazants** [comment]
@evolosen, is it still needed?

**Comment by ilya-lavrenov** [comment]
@andrei-kochin @rkazants is it still valid or we can close?

**Comment by rkazants** [comment]
> @andrei-kochin @rkazants is it still valid or we can close?

no, not needed. we can close it.
